Mathlib/Order/Normal.lean

```diff
@@ -0,0 +1,156 @@
+    cases hb.not_lt hb'
+  · intro b hb IH hab
+    obtain rfl | h := (lt_succ_iff_eq_or_lt_of_not_isMax hb).1 hab
+    · exact hs a
+    · exact (IH h).trans (hs b)
+  · intro b hb IH hab
+    have hab' := hb.succ_lt hab
+    exact (IH _ hab' (lt_succ_of_not_isMax hab.not_isMax)).trans_le
+      ((hl hb).1 (mem_image_of_mem _ hab'))
+
+theorem lt_iff_lt (hf : IsNormal f) : f a < f b ↔ a < b :=
+  hf.strictMono.lt_iff_lt
+
+theorem le_iff_le (hf : IsNormal f) : f a ≤ f b ↔ a ≤ b :=
+  hf.strictMono.le_iff_le
+
+theorem inj (hf : IsNormal f) : f a = f b ↔ a = b :=
+  hf.strictMono.injective.eq_iff
+
+theorem id_le {f : α → α} (hf : IsNormal f) [WellFoundedLT α] : id ≤ f :=
+  hf.strictMono.id_le
+
+theorem le_apply {f : α → α} (hf : IsNormal f) [WellFoundedLT α] : a ≤ f a :=
+  hf.strictMono.le_apply
+
+theorem isLUB_image_Iio_of_isSuccLimit (hf : IsNormal f) {a : α} (ha : IsSuccLimit a) :
+    IsLUB (f '' Iio a) (f a) := by
+  refine ⟨?_, mem_lowerBounds_upperBounds hf ha⟩
+  rintro _ ⟨b, hb, rfl⟩
+  exact (hf.strictMono hb).le
+
+theorem le_iff_forall_le (hf : IsNormal f) (ha : IsSuccLimit a) {b : β} :
+    f a ≤ b ↔ ∀ a' < a, f a' ≤ b := by
+  simpa [mem_upperBounds] using isLUB_le_iff (hf.isLUB_image_Iio_of_isSuccLimit ha)
+
+theorem lt_iff_exists_lt (hf : IsNormal f) (ha : IsSuccLimit a) {b : β} :
+    b < f a ↔ ∃ a' < a, b < f a' := by
+  simpa [mem_upperBounds] using lt_isLUB_iff (hf.isLUB_image_Iio_of_isSuccLimit ha)
+
+theorem map_isSuccLimit (hf : IsNormal f) (ha : IsSuccLimit a) : IsSuccLimit (f a) := by
+  refine ⟨?_, fun b hb ↦ ?_⟩
+  · obtain ⟨b, hb⟩ := not_isMin_iff.1 ha.not_isMin
+    exact not_isMin_iff.2 ⟨_, hf.strictMono hb⟩
+  · obtain ⟨c, hc, hc'⟩ := (hf.lt_iff_exists_lt ha).1 hb.lt
+    have hc' := hb.ge_of_gt hc'
+    rw [hf.le_iff_le] at hc'
+    exact hc.not_le hc'
+
+theorem map_isLUB (hf : IsNormal f) {s : Set α} (hs : IsLUB s a) (hs' : s.Nonempty) :
+    IsLUB (f '' s) (f a) := by
+  refine ⟨?_, fun b hb ↦ ?_⟩
+  · simp_rw [mem_upperBounds, forall_mem_image, hf.le_iff_le]
+    exact hs.1
+  · by_cases ha : a ∈ s
+    · simp_rw [mem_upperBounds, forall_mem_image] at hb
+      exact hb ha
+    · have ha' := hs.isSuccLimit_of_not_mem hs' ha
+      rw [le_iff_forall_le hf ha']
+      intro c hc
+      obtain ⟨d, hd, hcd, hda⟩ := hs.exists_between hc
+      simp_rw [mem_upperBounds, forall_mem_image] at hb
+      exact (hf.strictMono hcd).le.trans (hb hd)
+
+theorem _root_.InitialSeg.isNormal (f : α ≤i β) : IsNormal f where
+  strictMono := f.strictMono
+  mem_lowerBounds_upperBounds := by
+    intro a ha
+    rw [f.image_Iio]
+    exact (f.map_isSuccLimit ha).isLUB_Iio.2
+
+theorem _root_.PrincipalSeg.isNormal (f : α <i β) : IsNormal f :=
+  (f : α ≤i β).isNormal
+
+theorem _root_.OrderIso.isNormal (f : α ≃o β) : IsNormal f :=
+  f.toInitialSeg.isNormal
+
+protected theorem id : IsNormal (@id α) :=
+  (OrderIso.refl _).isNormal
+
+theorem trans (hg : IsNormal g) (hf : IsNormal f) : IsNormal (g ∘ f) where
```

Comment:
I think this would usually be called comp, since it's a composition for an unbundled property rather than a composition of a bundled one

Reply:
This matches the existing [`Ordinal.IsNormal.trans`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/SetTheory/Ordinal/Arithmetic.html#Ordinal.IsNormal.trans), but I do agree with you.